xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java

```diff
@@ -0,0 +1,183 @@
+          ADS_TYPE_URL_RDS, new HashMap<StreamObserver<DiscoveryResponse>, Set<String>>(),
+          ADS_TYPE_URL_CDS, new HashMap<StreamObserver<DiscoveryResponse>, Set<String>>(),
+          ADS_TYPE_URL_EDS, new HashMap<StreamObserver<DiscoveryResponse>, Set<String>>()
+          );
+  private final ImmutableMap<String, AtomicInteger> xdsVersions = ImmutableMap.of(
+      ADS_TYPE_URL_LDS, new AtomicInteger(1),
+      ADS_TYPE_URL_RDS, new AtomicInteger(1),
+      ADS_TYPE_URL_CDS, new AtomicInteger(1),
+      ADS_TYPE_URL_EDS, new AtomicInteger(1)
+  );
+  private final ImmutableMap<String, HashMap<StreamObserver<DiscoveryResponse>, AtomicInteger>>
+      xdsNonces = ImmutableMap.of(
+      ADS_TYPE_URL_LDS, new HashMap<StreamObserver<DiscoveryResponse>, AtomicInteger>(),
+      ADS_TYPE_URL_RDS, new HashMap<StreamObserver<DiscoveryResponse>, AtomicInteger>(),
+      ADS_TYPE_URL_CDS, new HashMap<StreamObserver<DiscoveryResponse>, AtomicInteger>(),
+      ADS_TYPE_URL_EDS, new HashMap<StreamObserver<DiscoveryResponse>, AtomicInteger>()
+  );
+
+
+  // treat all the resource types as state-of-the-world, send back all resources of a particular
+  // type when any of them change.
+  public <T extends Message> void setXdsConfig(final String type, final Map<String, T> resources) {
+    logger.log(Level.FINE, "setting config {0} {1}", new Object[]{type, resources});
+    syncContext.execute(new Runnable() {
+      @SuppressWarnings("unchecked")
+      @Override
+      public void run() {
+        HashMap<String, Message> copyResources =  new HashMap<>(resources);
+        xdsResources.put(type, copyResources);
+        String newVersionInfo = String.valueOf(xdsVersions.get(type).getAndDecrement());
+
+        for (Map.Entry<StreamObserver<DiscoveryResponse>, Set<String>> entry :
+            subscribers.get(type).entrySet()) {
+          DiscoveryResponse response = generateResponse(type, newVersionInfo,
+              String.valueOf(xdsNonces.get(type).get(entry.getKey()).incrementAndGet()),
+              entry.getValue());
+          entry.getKey().onNext(response);
+        }
+      }
+    });
+  }
+
+  @Override
+  public StreamObserver<DiscoveryRequest> streamAggregatedResources(
+      final StreamObserver<DiscoveryResponse> responseObserver) {
+    final StreamObserver<DiscoveryRequest> requestObserver =
+        new StreamObserver<DiscoveryRequest>() {
+      @Override
+      public void onNext(final DiscoveryRequest value) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            logger.log(Level.FINEST, "control plane received request {0}", value);
+            if (value.hasErrorDetail()) {
+              logger.log(Level.FINE, "control plane received nack resource {0}, error {1}",
+                  new Object[]{value.getResourceNamesList(), value.getErrorDetail()});
+              return;
+            }
+            if (value.getResourceNamesCount() <= 0) {
```

Comment:
This doesn't look right. It should update requestedResourceNames. That can be done by either deleting this block and provide the client with zero entries, or it should be handled via `subscribers.get(type).remove(responseObserver)`.